### PR TITLE
Improve Documentation #2845 - lexical/utils

### PR DESCRIPTION
**Pull request by HarrySIV**
With Acy's help I've tried to follow his example and add inline api documentation, though I might have gone a bit overboard. Please let me know if there's anything I can do to improve the PR.

### CODE REVIEW

**Comment by HarrySIV** [comment]
I'm really unsure as to why the e2e tests are failing. I wonder if using typedoc inline comments inside another one caused an issue, but I don't really know.

**Comment by acywatson** [comment]
e2es are just being flaky

the vercel deployment is failing because of this:

```


Expand 62 Lines
--
17:33:57.095 | SyntaxError: /vercel/path0/packages/lexical-website/docs/api/modules/lexical_utils.md: Expected corresponding JSX closing tag for <div>. (538:70)
17:33:57.095 | 536 \| eg. element = `}<div class="element-inner active small">{`
17:33:57.096 | 537 \| removeClassNamesFromElement(element, `}{`['active small', true, null]`}{`);
```

File: packages/lexical-utils/src/index.ts [hunk 1]
```diff
@@ -226,22 +285,45 @@ export function $findMatchingParent(
 
 type Func = () => void;
 
+/**
+ * Executes each function passed into the mergeRegister function argument.
```
Comment (acywatson) [inline packages/lexical-utils/src/index.ts:289]: Almost. What this actually does is returns a function that will execute all of the functions passed to mergeRegister when called. The purpose is to make it easier to register a bunch of lexical listeners and then tear them down with a single function call. For example, in a React useEffect.

This one could potentially benefit from an @example annotation

File: packages/lexical-utils/src/index.ts [hunk 2]
```diff
@@ -226,22 +285,45 @@ export function $findMatchingParent(
...
+ * @returns undefined
+ */
 export function mergeRegister(...func: Array<Func>): () => void {
   return () => {
     func.forEach((f) => f());
   };
 }
 
+/**
+ *
```
Comment (acywatson) [inline packages/lexical-utils/src/index.ts:300]: This attempts to resolve nested element nodes of the same type into a single node of that type. used mostly with marks/commenting

File: packages/lexical-utils/src/index.ts [hunk 3]
```diff
@@ -226,22 +285,45 @@ export function $findMatchingParent(
...
+  /**
+   * @param node - node to be tested against the targetNode instance
+   * @returns true if there is a match, false otherwise
+   */
   const $isTargetNode = (node: LexicalNode | null | undefined): node is N => {
     return node instanceof targetNode;
   };
 
+  /**
+   * Looks for an ancestor of a node of type targetNode. Ensures there are not any children of the node of the same targetNode instance
```
Comment (acywatson) [inline packages/lexical-utils/src/index.ts:322]: remove

File: packages/lexical-utils/src/index.ts [hunk 4]
```diff
@@ -270,6 +352,10 @@ export function registerNestedElementResolver<N extends ElementNode>(
     return null;
   };
 
+  /**
```
Comment (acywatson) [inline packages/lexical-utils/src/index.ts:355]: remove

File: packages/lexical-utils/src/index.ts [hunk 5]
```diff
@@ -307,6 +393,12 @@ export function registerNestedElementResolver<N extends ElementNode>(
   return editor.registerNodeTransform(targetNode, elementNodeTransform);
 }
 
+/**
+ * Clones the editor and marks it as dirty to be reconciled. If there was a selection,
```
Comment (acywatson) [inline packages/lexical-utils/src/index.ts:397]: I wonder how this is different from setEditorState... We can leave as-is for now and polish it up later.

File: packages/lexical-utils/src/index.ts (hunk 1 above)
Comment (HarrySIV) [reply packages/lexical-utils/src/index.ts]: This took me a minute to wrap my head around, but I think I understand how it works especially with useEffect. I hope my explanation is clear and helpful.

**Comment by acywatson** [comment]
Hmm - looks like a prettier error in there somewhere. Can you run prettier locally to see if you can figure out what the issue is?

**Comment by HarrySIV** [comment]
Looked like some extra whitespace at the end of some sentences. Sorry...

**Review by acywatson — APPROVED** [review]

**Pull request merged**